components/org.wso2.carbon.uis/src/main/java/org/wso2/carbon/uis/internal/io/deployment/ArtifactAppDeployer.java

```diff
@@ -27,125 +27,170 @@
     private static final Logger LOGGER = LoggerFactory.getLogger(ArtifactAppDeployer.class);
 
-    private final Path appsRepository;
-    private final Set<AppDeploymentEventListener> eventListeners = new HashSet<>();
-    private final Set<String> deployedAppNames = new HashSet<>();
+    private final ArtifactType<String> artifactType;
+    private final URL deploymentLocation;
+    private final ConcurrentMap<String, App> deployedApps;
+    private AppDeploymentEventListener appDeploymentEventListener;
 
     /**
-     * Creates a new app deployer that locates apps from {@code <CARBON_HOME>/deployment/reactapps} directory.
+     * Creates a new app deployer.
      */
     public ArtifactAppDeployer() {
-        this(Paths.get(System.getProperty("carbon.home", "."), "deployment", "reactapps"));
-    }
-
-    /**
-     * Creates a new app deployer that locates apps from the given directory.
-     *
-     * @param appsRepository app repository directory
-     */
-    public ArtifactAppDeployer(Path appsRepository) {
-        this.appsRepository = appsRepository;
+        this.artifactType = new ArtifactType<>(ARTIFACT_TYPE);
+        URL deploymentLocationUrl = null;
+        try {
+            deploymentLocationUrl = new URL(DEPLOYMENT_LOCATION);
+        } catch (MalformedURLException e) {
+            LOGGER.error("Invalid URL '{}' as app deployment location.", DEPLOYMENT_LOCATION);
+        }
+        this.deploymentLocation = deploymentLocationUrl;
+        this.deployedApps = new ConcurrentHashMap<>();
     }
 
     @Reference(name = "deploymentListener",
                service = AppDeploymentEventListener.class,
-               cardinality = ReferenceCardinality.AT_LEAST_ONE,
+               cardinality = ReferenceCardinality.MANDATORY,
                policy = ReferencePolicy.DYNAMIC,
                unbind = "unregisterListener")
     protected void registerListener(AppDeploymentEventListener listener) {
-        this.eventListeners.add(listener);
+        this.appDeploymentEventListener = listener;
         LOGGER.debug("An instance of class '{}' registered as an app deployment listener.",
                      listener.getClass().getName());
     }
 
     protected void unregisterListener(AppDeploymentEventListener listener) {
-        this.eventListeners.remove(listener);
+        this.appDeploymentEventListener = null;
         LOGGER.debug("An instance of class '{}' unregistered as an app deployment listener.",
                      listener.getClass().getName());
     }
 
     @Activate
     protected void activate(BundleContext bundleContext) {
-        start();
         LOGGER.debug("Carbon UI server app deployer activated.");
     }
 
     @Deactivate
     protected void deactivate(BundleContext bundleContext) {
-        stop();
+        try {
+            for (String artifactKey : deployedApps.keySet()) {
+                undeploy(artifactKey);
+            }
+        } catch (CarbonDeploymentException e) {
```

Comment:
If one artifact gets failed to undeploy, all others will not get undeployed.

Reply:
fixed in commit 75af34228af9dbc1697a74f7e3c9e600412f04ff